### PR TITLE
E2E: beacon APIs Part 1

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

E2E

**What does this PR do? Why is it needed?**

Expanding on the E2E with tests for beacon APIs

### CODE REVIEW

File: testing/endtoend/components/lighthouse_beacon.go [hunk 1]
```diff
@@ -182,7 +182,7 @@ func (node *LighthouseBeaconNode) Start(ctx context.Context) error {
 		fmt.Sprintf("--enr-tcp-port=%d", e2e.TestParams.Ports.LighthouseBeaconNodeP2PPort+index),
 		fmt.Sprintf("--port=%d", e2e.TestParams.Ports.LighthouseBeaconNodeP2PPort+index),
 		fmt.Sprintf("--http-port=%d", e2e.TestParams.Ports.LighthouseBeaconNodeHTTPPort+index),
-		fmt.Sprintf("--target-peers=%d", 10),
+		fmt.Sprintf("--target-peers=%d", e2e.TestParams.LighthouseBeaconNodeCount),
```
Comment (nisdas) [inline testing/endtoend/components/lighthouse_beacon.go:185]: Why was this changed ?

File: testing/endtoend/evaluators/beaconapi_evaluators/types.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package beaconapi_evaluators
+
+type validatorJson struct {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply testing/endtoend/evaluators/beaconapi_evaluators/types.go:3]: will remove some of these, was mainly focused on beacon_api

File: testing/endtoend/components/lighthouse_beacon.go (hunk 1 above)
Comment (james-prysm) [reply testing/endtoend/components/lighthouse_beacon.go:185]: It errors on peer discovery, so I thought it should be the amount we choose

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api_verify.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	beacon := []apiComparisonFunc{
+		withCompareBeaconAPIs,
+	}
+	for beaconNodeIdx, conn := range conns {
+		if err := runAPIComparisonFunctions(
+			beaconNodeIdx,
+			conn,
+			beacon...,
+		); err != nil {
+			return errors.Wrap(err, "beaconAPI verify")
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api_verify.go:34]: the wrapping error is a bit meaningless

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api_verify.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		}
+	}
+	return nil
+}
+
+func runAPIComparisonFunctions(beaconNodeIdx int, conn *grpc.ClientConn, fs ...apiComparisonFunc) error {
+	for _, f := range fs {
+		if err := f(beaconNodeIdx, conn); err != nil {
+			return errors.Wrap(err, "runAPIComparisonFunctions")
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api_verify.go:43]: the wrapping error is a bit meaningless

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+	}
+	currentEpoch := slots.EpochsSinceGenesis(genesisData.Data.GenesisTime.AsTime())
+
+	for path, meta := range beaconPathsAndObjects {
+		for key := range meta.prysmResps {
+			switch key {
+			case "json":
+				jsonparams := meta.params("json", currentEpoch)
+				if len(jsonparams) == 0 {
+					continue
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:238]: I am not exactly sure what this signifies, but shouldn't this be an error condition? Similarly for ssz.

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+				if err := compareJSONMulticlient(beaconNodeIdx,
+					meta.basepath,
+					apipath,
+					beaconPathsAndObjects[path].prysmResps[key],
+					beaconPathsAndObjects[path].lighthouseResps[key],
+					meta.customEvaluation,
+				); err != nil {
+					return err
+				}
+				fmt.Printf("prysm ob: %v/n", beaconPathsAndObjects[path].prysmResps[key])
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:251]:
```suggestion
				fmt.Printf("prysm object: %v/n", beaconPathsAndObjects[path].prysmResps[key])
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+				fmt.Printf("lighthouse ob: %v", beaconPathsAndObjects[path].prysmResps[key])
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:252]:
```suggestion
				fmt.Printf("lighthouse object: %v", beaconPathsAndObjects[path].lighthouseResps[key])
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+					continue
+				}
+				apipath := pathFromParams(path, sszparams)
+				fmt.Printf("ssz api path: %s", apipath)
+				prysmr, lighthouser, err := compareSSZMulticlient(beaconNodeIdx, meta.basepath, apipath)
+				if err != nil {
+					return err
+				}
+				beaconPathsAndObjects[path].prysmResps[key] = prysmr
+				beaconPathsAndObjects[path].lighthouseResps[key] = lighthouser
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:265]: Is this needed?

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+		return errors.New("failed to cast type")
+	}
+	lighthouseForkData, ok := forkPathData.lighthouseResps["json"].(*apimiddleware.StateForkResponseJson)
+	if !ok {
+		return errors.New("failed to cast type")
+	}
+	if prysmForkData.Data.Epoch != lighthouseForkData.Data.Epoch {
+		return fmt.Errorf("prysm response %v does not match lighthouse response %v",
+			prysmForkData,
+			lighthouseForkData)
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:283]:
```suggestion
		return fmt.Errorf("prysm epoch %v does not match lighthouse epoch %v",
			prysmForkData.Data.Epoch,
			lighthouseForkData.Data.Epoch)
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+		return errors.New("failed to cast type")
+	}
+	if finalizedEpoch < 8 {
+		blockP := &ethpb.SignedBeaconBlock{}
+		blockL := &ethpb.SignedBeaconBlock{}
+		if err := blockL.UnmarshalSSZ(sszrspL); err != nil {
+			return errors.Wrap(err, "lighthouse ssz error")
+		}
+		if err := blockP.UnmarshalSSZ(sszrspP); err != nil {
+			return errors.Wrap(err, "prysm ssz error")
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:306]:
```suggestion
			return errors.Wrap(err, "failed to unmarshal Prysm block")
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+		}
+		if len(blockP.Signature) == 0 || len(blockL.Signature) == 0 || hexutil.Encode(blockP.Signature) != hexutil.Encode(blockL.Signature) {
+			return fmt.Errorf("prysm response %v does not match lighthouse response %v",
+				blockP,
+				blockL)
+		}
+	} else if finalizedEpoch > 8 && finalizedEpoch < 10 {
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:313]:
```suggestion
	} else if finalizedEpoch >= helpers.AltairE2EForkEpoch + 2 && finalizedEpoch < helpers.BellatrixE2EForkEpoch {
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+		}
+		if err := blockP.UnmarshalSSZ(sszrspP); err != nil {
+			return errors.Wrap(err, "prysm ssz error")
+		}
+
+		if len(blockP.Signature) == 0 || len(blockL.Signature) == 0 || hexutil.Encode(blockP.Signature) != hexutil.Encode(blockL.Signature) {
+			return fmt.Errorf("prysm response %v does not match lighthouse response %v",
+				blockP,
+				blockL)
+		}
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:327]: This block of code is duplicated here and for Bellatrix fork. I think you should be able to deduplicate it.

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+}
+
+func compareJSONMulticlient(beaconNodeIdx int, base string, path string, respJSONPrysm interface{}, respJSONLighthouse interface{}, customEvaluator func(interface{}, interface{}) error) error {
+	if err := doMiddlewareJSONGetRequest(
+		base,
+		path,
+		beaconNodeIdx,
+		respJSONPrysm,
+	); err != nil {
+		return errors.Wrap(err, "prysm json error")
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:369]:
```suggestion
		return errors.Wrap(err, "could not perform GET request for Prysm JSON")
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 14]
```diff
@@ -0,0 +1,440 @@
...
+	}
+
+	if err := doMiddlewareJSONGetRequest(
+		base,
+		path,
+		beaconNodeIdx,
+		respJSONLighthouse,
+		"lighthouse",
+	); err != nil {
+		return errors.Wrap(err, "lighthouse json error")
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:379]:
```suggestion
		return errors.Wrap(err, "could not perform GET request for Lighthouse JSON")
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 15]
```diff
@@ -0,0 +1,440 @@
...
+
+func compareSSZMulticlient(beaconNodeIdx int, base string, path string) ([]byte, []byte, error) {
+	sszrspL, err := doMiddlewareSSZGetRequest(
+		base,
+		path,
+		beaconNodeIdx,
+		"lighthouse",
+	)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "lighthouse json error")
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:396]:
```suggestion
		return nil, nil, errors.Wrap(err, "could not perform GET request for Lighthouse SSZ")
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 16]
```diff
@@ -0,0 +1,440 @@
...
+	}
+
+	sszrspP, err := doMiddlewareSSZGetRequest(
+		base,
+		path,
+		beaconNodeIdx,
+	)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "prysm json error")
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:405]:
```suggestion
		return nil, nil, errors.Wrap(err, "could not perform GET request for Prysm SSZ")
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 17]
```diff
@@ -0,0 +1,440 @@
...
+	}
+	if !bytes.Equal(sszrspL, sszrspP) {
+		return nil, nil, fmt.Errorf("prysm ssz response %s does not match lighthouse ssz response %s",
+			hexutil.Encode(sszrspP),
+			hexutil.Encode(sszrspL))
+	}
+	fmt.Printf("prysm ssz: %v", sszrspP)
+	fmt.Printf("lighthouse ssz: %v", sszrspL)
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:413]: I would personally not print SSZ values here because they can be huge for state objects.

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 18]
```diff
@@ -0,0 +1,440 @@
...
+	return sszrspP, sszrspL, nil
+}
+
+func compareResponseObjects(prysmResp interface{}, lighthouseResp interface{}) error {
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:417]: Maybe change the name to `compareJSONObjects` to inform this should not be used for other encodings

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 19]
```diff
@@ -0,0 +1,440 @@
...
+	if !reflect.DeepEqual(prysmResp, lighthouseResp) {
+		p, err := json.Marshal(prysmResp)
+		if err != nil {
+			return errors.Wrap(err, "prysm json")
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:421]:
```suggestion
			return errors.Wrap(err, "failed to marshal Prysm response to JSON")
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 20]
```diff
@@ -0,0 +1,440 @@
...
+		}
+		l, err := json.Marshal(lighthouseResp)
+		if err != nil {
+			return errors.Wrap(err, "lighthouse json")
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:425]:
```suggestion
			return errors.Wrap(err, "failed to marshal Lighthouse response to JSON")
```

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 21]
```diff
@@ -0,0 +1,440 @@
...
+		}
+		return fmt.Errorf("prysm response %s does not match lighthouse response %s",
+			string(p),
+			string(l))
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:429]: I would be careful with printing whole objects because they can be huge for states.

**Review by rkapka — CHANGES_REQUESTED** [review]
Requesting changes because of commented out code.

File: testing/endtoend/mainnet_scenario_e2e_test.go [hunk 22]
```diff
@@ -6,13 +6,17 @@ import (
...
+//	runner.config.Evaluators = scenarioEvalsMulti()
+//	runner.config.EvalInterceptor = runner.multiScenarioMulticlient
+//	runner.scenarioRunner()
+//}
 
-func TestEndToEnd_MultiScenarioRun_Multiclient(t *testing.T) {
-	runner := e2eMainnet(t, false /*usePrysmSh*/, true /*useMultiClient*/, types.WithEpochs(22))
-	runner.config.Evaluators = scenarioEvalsMulti()
-	runner.config.EvalInterceptor = runner.multiScenarioMulticlient
-	runner.scenarioRunner()
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply testing/endtoend/mainnet_scenario_e2e_test.go:17]: it's not it's commented out for testing will uncomment

File: testing/endtoend/evaluators/node.go [hunk 23]
```diff
@@ -101,6 +101,7 @@ func peersConnect(conns ...*grpc.ClientConn) error {
 	ctx := context.Background()
 	for _, conn := range conns {
 		nodeClient := eth.NewNodeClient(conn)
+		time.Sleep(connTimeDelay * 3)
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply testing/endtoend/evaluators/node.go:104]: originally I was trying to see if it's just latency on evaluators for the peer to peer flaking issues I'll remove this now as I don't think it resolves them

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go (hunk 21 above)
Comment (james-prysm) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:424]: I think it's helpful if it's erroring out. ideally it doesn't error out, but hopefully if it does we can try to find out why. maybe I should add a debug log?

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api_verify.go (hunk 4 above)
Comment (james-prysm) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api_verify.go:43]: was originally trying to track where some of the errors are coming from rolling this back

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go (hunk 5 above)
Comment (james-prysm) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:235]: no this decides what params go into the url, and skips if there's nothing in the returned. it helps construct the url with some value and so I used skip but I could also use empty string for the ones that should compare some how but don't have parameter values.

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api_verify.go (hunk 3 above)
Comment (james-prysm) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api_verify.go:34]: will remove again was trying to pinpoint some earlier errors. where it was bubbling from.

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 24]
```diff
@@ -0,0 +1,440 @@
...
+	prysmResps       map[string]interface{}
+	lighthouseResps  map[string]interface{}
+	customEvaluation func(interface{}, interface{}) error
+}
+
+var beaconPathsAndObjects = map[string]metadata{
+	"/beacon/genesis": {
+		basepath: v1MiddlewarePathTemplate,
+		params: func(_ string, _ types.Epoch) []string {
+			return []string{"skip"}
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:39]: would result in skipped comparison , explained in a later comment

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 25]
```diff
@@ -0,0 +1,440 @@
...
+			"json": &apimiddleware.BlockHeaderResponseJson{},
+		},
+		lighthouseResps: map[string]interface{}{
+			"json": &apimiddleware.BlockHeaderResponseJson{},
+		},
+	},
+	"/node/identity": {
+		basepath: v1MiddlewarePathTemplate,
+		params: func(_ string, _ types.Epoch) []string {
+			return []string{"skip"}
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:163]: would result in skipped comparison , explained in a later comment

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 26]
```diff
@@ -0,0 +1,440 @@
...
+			if castedl.Data == nil {
+				return errors.New("lighthouse node identity was empty")
+			}
+			return nil
+		},
+	},
+	"/node/peers": {
+		basepath: v1MiddlewarePathTemplate,
+		params: func(_ string, _ types.Epoch) []string {
+			return []string{"skip"}
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:192]: would result in skipped comparison , explained in a later comment

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go (hunk 8 above)
Comment (james-prysm) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:262]: yeah it's later used in the unmarshalling and comparing the unmarshals

File: testing/endtoend/mainnet_scenario_e2e_test.go [hunk 27]
```diff
@@ -6,13 +6,17 @@ import (
...
+//func TestEndToEnd_MainnetConfig_MultiClient(t *testing.T) {
+//	e2eMainnet(t, false /*usePrysmSh*/, true /*useMultiClient*/).run()
+//}
+//
+//func TestEndToEnd_MultiScenarioRun_Multiclient(t *testing.T) {
+//	runner := e2eMainnet(t, false /*usePrysmSh*/, true /*useMultiClient*/, types.WithEpochs(22))
+//	runner.config.Evaluators = scenarioEvalsMulti()
+//	runner.config.EvalInterceptor = runner.multiScenarioMulticlient
+//	runner.scenarioRunner()
+//}
```
Comment (rkapka) [inline testing/endtoend/mainnet_scenario_e2e_test.go:18]: Commented out code

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go (hunk 5 above)
Comment (rkapka) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:235]: Just return `[]string{}` instead. In such case the `for` loop inside `pathFromParams` will never run.

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go [hunk 28]
```diff
@@ -0,0 +1,433 @@
...
+}
+
+func pathFromParams(path string, params []string) string {
+	var apiPath string
+	for index := range params {
+		apiPath = strings.Replace(path, fmt.Sprintf("{param%d}", index+1), params[index], 1)
+	}
+	if apiPath == "" {
+		return path
+	}
```
Comment (rkapka) [inline testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:431]: Shouldn't this be an error condition? It seems to me that someone passed in a malformed path.

File: testing/endtoend/endtoend_setup_test.go [hunk 29]
```diff
@@ -115,16 +116,17 @@ func e2eMainnet(t *testing.T, usePrysmSh, useMultiClient bool, cfgo ...types.E2E
 		ev.PeersConnect,
 		ev.HealthzCheck,
 		ev.MetricsCheck,
-		ev.ValidatorsAreActive,
```
Comment (rkapka) [inline testing/endtoend/endtoend_setup_test.go:118]: Should this be removed?

File: testing/endtoend/mainnet_scenario_e2e_test.go [hunk 30]
```diff
@@ -1,18 +1,21 @@
...
+func TestEndToEnd_MainnetConfig_Multiclient_CrossClient(t *testing.T) {
+	e2eMainnet(t, false /*usePrysmSh*/, true /*useMultiClient*/, types.WithValidatorCrossClient()).run()
 }
+
+//func TestEndToEnd_MultiScenarioRun_Multiclient(t *testing.T) {
+//	runner := e2eMainnet(t, false /*usePrysmSh*/, true /*useMultiClient*/, types.WithEpochs(22))
+//	runner.config.Evaluators = scenarioEvalsMulti()
+//	runner.config.EvalInterceptor = runner.multiScenarioMulticlient
+//	runner.scenarioRunner()
+//}
```
Comment (rkapka) [inline testing/endtoend/mainnet_scenario_e2e_test.go:21]: Commented code

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go (hunk 28 above)
Comment (james-prysm) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:431]: It should return the input if there are no parameters to replace. previously it just returned the api Path. the call will fail but that shouldn't mean this function should error.

File: testing/endtoend/endtoend_setup_test.go (hunk 29 above)
Comment (james-prysm) [reply testing/endtoend/endtoend_setup_test.go:118]: This is removed in favor of Kasey's new DepositedValidatorsAreActive evaluator.

File: testing/endtoend/mainnet_scenario_e2e_test.go (hunk 30 above)
Comment (james-prysm) [reply testing/endtoend/mainnet_scenario_e2e_test.go:21]: I was still testing lol will be uncommented soon

File: testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go (hunk 28 above)
Comment (rkapka) [reply testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go:431]: Ah, right. In that case I think you can make this code a little simpler by declaring `apiPath := path` first, and then simply returning `apiPath` after the loop.

**Review by rkapka — APPROVED** [review]
